### PR TITLE
Add Redis maintenace job

### PR DESCRIPTION
**Pull request by zugao**
## Summary

* This PR adds Redis maintenance job. The idea is get the latest patch version a Redis instance and update the Helm Release object.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.

### CODE REVIEW

File: pkg/maintenance/redis.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+		return fmt.Errorf("cannot patch release %s: %v", release.Name, err)
+	}
+	return nil
+}
+
+func (r *Redis) patchRelease(ctx context.Context, version semver.Version, release *v1beta1.Release) error {
+	patchOps := []postgres.Jsonpatch{
+		{
+			Op:    "replace",
+			Path:  "/spec/forProvider/values/image/tag",
```
Comment (Kidswiss) [inline pkg/maintenance/redis.go:81]: This gets overwrite again by the P+T composition: https://github.com/vshn/component-appcat/blob/master/component/vshn_redis.jsonnet#L514

We'll have to handle this via comp-function.

**Review by Kidswiss — CHANGES_REQUESTED** [review]

File: pkg/maintenance/redis.go (hunk 1 above)
Comment (zugao) [reply pkg/maintenance/redis.go:81]: As agreed I added a new comp function that will update the helm values first.

**Review by Kidswiss — APPROVED** [review]
LGTM

File: pkg/comp-functions/functions/vshnredis/maintenance.go [hunk 2]
```diff
@@ -9,9 +9,8 @@ import (
 )
 
 var (
-	service         = "redis"
-	maintRolename   = "crossplane:appcat:job:redis:maintenance"
-	maintSecretName = "maintenancesecret"
+	service          = "redis"
+	helmBasedService = true
 
 	policyRules = []rbacv1.PolicyRule{
```
Comment (Kidswiss) [inline pkg/comp-functions/functions/vshnredis/maintenance.go:15]: This is probably not necessary anymore?

**Review by Kidswiss — CHANGES_REQUESTED** [review]
A few things that are confusing, the rest LGTM

File: pkg/comp-functions/functions/vshnredis/maintenance.go [hunk 3]
```diff
@@ -43,6 +42,6 @@ func AddMaintenanceJob(ctx context.Context, iof *runtime.Runtime) runtime.Result
 	instanceNamespace := getInstanceNamespace(comp)
 	schedule := comp.Spec.Parameters.Maintenance
 
-	return maintenance.New(comp, iof, schedule, policyRules, instanceNamespace, maintRolename, service).
+	return maintenance.New(comp, iof, schedule, policyRules, helmBasedService, instanceNamespace, service).
```
[the remark this answers is not in the document]
Comment (zugao) [reply pkg/comp-functions/functions/vshnredis/maintenance.go:45]: yeah I though about it though I went with the constructor as to emphasize the importance of this value

File: pkg/comp-functions/functions/common/maintenance/maintenance.go [hunk 4]
```diff
@@ -59,15 +61,15 @@ var (
 )
 
 // New creates a Maintenance object with required attributes
-func New(r client.Object, iof *runtime.Runtime, sc vshnv1.VSHNDBaaSMaintenanceScheduleSpec, pr []rbacv1.PolicyRule, in, mr, s string) *Maintenance {
+func New(r client.Object, iof *runtime.Runtime, sc vshnv1.VSHNDBaaSMaintenanceScheduleSpec, pr []rbacv1.PolicyRule, h bool, in, s string) *Maintenance {
```
[the remark this answers is not in the document]
Comment (zugao) [reply pkg/comp-functions/functions/common/maintenance/maintenance.go:64]: you can always check the code? I describe it pretty well each field in the struct

[the remark this answers is not in the document]
Comment (Kidswiss) [reply pkg/comp-functions/functions/common/maintenance/maintenance.go:64]: Yeah, but having to go into the code and look at it is not a very good dev experience. Especially when we should consider making this more approachable for people outside of Schedar.

File: pkg/comp-functions/functions/vshnredis/maintenance.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnredis/maintenance.go:45]: But then we should make the `policyRules` optional, as they are not always needed in the constructor.

File: pkg/comp-functions/functions/vshnredis/maintenance.go (hunk 2 above)
Comment (zugao) [reply pkg/comp-functions/functions/vshnredis/maintenance.go:15]: cleaned the bool variable though the service is still necessary

File: pkg/comp-functions/functions/vshnredis/release.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+			},
+		},
+	}
+
+	masterMap["extraVolumes"] = volumes
+
+	volumeMounts := []corev1.VolumeMount{
+		{
+			Name:      backupScriptCMName,
+			MountPath: "/scripts",
```
Comment (wejdross) [inline pkg/comp-functions/functions/vshnredis/release.go:160]: is it some general VSHN script or is it under our control?

File: pkg/comp-functions/functions/common/maintenance/maintenance.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (zugao) [reply pkg/comp-functions/functions/common/maintenance/maintenance.go]: sure I updated it ;)

File: pkg/comp-functions/functions/vshnredis/release.go (hunk 5 above)
Comment (zugao) [reply pkg/comp-functions/functions/vshnredis/release.go:160]: This should be under out control. @Kidswiss should have more to say about that.

Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnredis/release.go:160]: It's our script, see: https://github.com/vshn/appcat/blob/master/pkg/comp-functions/functions/vshnredis/script/backup.sh

File: pkg/comp-functions/functions/vshnredis/maintenance.go (hunk 2 above)
Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnredis/maintenance.go:13]: I wasn't talking about either the bool nor the service but about the `policyRules`. The clusterRole is deployed via the component, right? So why do we have the same role here again?

Comment (zugao) [reply pkg/comp-functions/functions/vshnredis/maintenance.go:13]: Ah yes could catch, let me update it :)

**Review by Kidswiss — APPROVED** [review]

**Pull request merged**